### ISSUE TITLE
The values of variable loss2 always are zero

### ISSUE THREAD

**Issue by SidneyLann** [deeplearning4j/deeplearning4j#8542]
#### Issue Description

Please describe our issue, along with:
- expected behavior
The value of variable loss2 should not be zero.

- encountered behavior
The values of variable loss2 always are zero.

Build model...
[main] INFO org.nd4j.linalg.factory.Nd4jBackend - Loaded [CpuBackend] backend
[main] INFO org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for linear algebra: 4
[main] INFO org.nd4j.nativeblas.Nd4jBlas - Number of threads used for OpenMP BLAS: 4
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Backend used: [CPU]; OS: [Linux]
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Cores: [4]; Memory: [1.8GB];
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Blas vendor: [OPENBLAS]
16 Dec, 2019 17:53:43  io.vertx.core.spi.resolver.ResolverProvider
INFO: Using the default address resolver as the dns resolver could not be loaded
[vert.x-eventloop-thread-0] INFO org.deeplearning4j.ui.VertxUIServer - Deeplearning4j UI server started at: http://localhost:9000
[main] INFO org.deeplearning4j.ui.VertxUIServer - Deeplearning4j UI server started at: http://localhost:9000
[AMDSI prefetch thread] INFO org.nd4j.autodiff.samediff.SameDiff - Creating new InferenceSession for thread 18
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 0, iteration 10: 0.00552
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 0, iteration 20: 0.01176, batches/sec: 0.00780, examples/sec: 0.99875
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 0, iteration 30: 0.01545, batches/sec: 0.00784, examples/sec: 1.00299
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 0, iteration 40: 0.01692, batches/sec: 0.00784, examples/sec: 1.00311
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 0, iteration 50: 0.01755, batches/sec: 0.00780, examples/sec: 0.99861
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 0, iteration 60: 0.01812, batches/sec: 0.00781, examples/sec: 1.00010
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 0, iteration 70: 0.01883, batches/sec: 0.00787, examples/sec: 1.00673
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Epoch 0 complete on iteration 73 - 74 batches (9379 examples) in 25.66 hr - 8.01E-3 batches/sec, 1.02E0 examples/sec, 0 ms ETL time
[main] INFO org.nd4j.autodiff.listeners.checkpoint.CheckpointListener - Model checkpoint saved: epoch 0, iteration 73, path: /home/sidney/ner/model/SameDiff_checkpoint-0_epoch-0_iter-73_2019-12-16_22-29-58.bin
[main] INFO org.nd4j.autodiff.samediff.SameDiff - Creating new InferenceSession for thread 1
[main] INFO org.nd4j.autodiff.samediff.SameDiff - Creating new InferenceSession for thread 1
[AMDSI prefetch thread] INFO org.nd4j.autodiff.samediff.SameDiff - Creating new InferenceSession for thread 22
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 1, iteration 80: 0.01964(1 ms ETL in 10 iter)
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 1, iteration 90: 0.02058, batches/sec: 0.00840, examples/sec: 0.99765
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 1, iteration 100: 0.02168, batches/sec: 0.00771, examples/sec: 0.98740
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 1, iteration 110: 0.02274, batches/sec: 0.00775, examples/sec: 0.99206
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 1, iteration 120: 0.02381, batches/sec: 0.00769, examples/sec: 0.98482
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 1, iteration 130: 0.02490, batches/sec: 0.00771, examples/sec: 0.98704
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 1, iteration 140: 0.02579, batches/sec: 0.00777, examples/sec: 0.99444
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Epoch 1 complete on iteration 147 - 74 batches (9379 examples) in 2.63E1 hr - 7.82E-3 batches/sec, 9.91E-1 examples/sec, 1 ms ETL time(1.06E-5 %)
[main] INFO org.nd4j.autodiff.listeners.checkpoint.CheckpointListener - Model checkpoint saved: epoch 1, iteration 147, path: /home/sidney/ner/model/SameDiff_checkpoint-1_epoch-1_iter-147_2019-12-17_01-09-58.bin
[main] INFO org.nd4j.autodiff.samediff.SameDiff - Creating new InferenceSession for thread 1
[AMDSI prefetch thread] INFO org.nd4j.autodiff.samediff.SameDiff - Creating new InferenceSession for thread 23
[main] INFO org.nd4j.autodiff.listeners.impl.ScoreListener - Loss at epoch 2, iteration 150: 0.02663

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version: beta6
* Platform information (OS, etc)
* CUDA version, if used
* NVIDIA driver version, if in use

#### Additional Information
package com.pcng.ner.bert;

import java.io.File;
import java.util.List;

import org.deeplearning4j.ui.VertxUIServer;
import org.nd4j.autodiff.listeners.checkpoint.CheckpointListener;
import org.nd4j.autodiff.listeners.impl.ScoreListener;
import org.nd4j.autodiff.listeners.impl.UIListener;
import org.nd4j.autodiff.listeners.records.History;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.autodiff.samediff.TrainingConfig;
import org.nd4j.evaluation.classification.Evaluation;
import org.nd4j.evaluation.classification.Evaluation.Metric;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.learning.config.Adam;
import org.nd4j.weightinit.impl.XavierInitScheme;

public class NerBertTraining {

  public static void main(String[] args) {
    try {
      int batchSize = args.length > 0 ? Integer.valueOf(args[0]) : 1;
      int nEpochs = args.length > 0 ? Integer.valueOf(args[1]) : 1;
      NerUtil.MINI_BATCH_SIZE = batchSize;

      SameDiff sd;
      File modelFolder = new File(NerUtil.DATA_HOME + "/ner/model");

      if (modelFolder.list().length > 0) {
        System.out.println("Load model...");
        sd = CheckpointListener.loadLastCheckpoint(modelFolder, true);
      } else {
        System.out.println("Build model...");
        sd = SameDiff.create();

        SDVariable input2 = sd.placeHolder("bertOutput", DataType.FLOAT, -1, NerUtil.MAX_SENTENCE_LENGTH, NerUtil.MAX_SENTENCE_LENGTH * 6);
        SDVariable label = sd.placeHolder("label", DataType.FLOAT, -1, NerUtil.MAX_SENTENCE_LENGTH, NerUtil.getLabelNum());
        SDVariable wOut = sd.var("wOut", new XavierInitScheme('c', NerUtil.MAX_SENTENCE_LENGTH * NerUtil.MAX_SENTENCE_LENGTH * 6, NerUtil.MAX_SENTENCE_LENGTH * NerUtil.getLabelNum()), DataType.FLOAT,
            NerUtil.MAX_SENTENCE_LENGTH, NerUtil.MAX_SENTENCE_LENGTH * 6, NerUtil.getLabelNum());
        SDVariable bOut = sd.one("bOut", DataType.FLOAT, NerUtil.getLabelNum());
        SDVariable permute1 = sd.permute("permute1", input2, 1, 0, 2);
        SDVariable multiply4Output = sd.mmul("multiply4Output", permute1, wOut);
        SDVariable permute2 = sd.permute("permute2", multiply4Output, 1, 0, 2);
        SDVariable output2 = sd.nn.biasAdd("output2", permute2, bOut, false);
        SDVariable loss = sd.loss().softmaxCrossEntropy("loss2", label, output2);

        sd.setLossVariables(loss);
        sd.setOutputs("output2");

        double learningRate = 1e-3;
        TrainingConfig config = new TrainingConfig.Builder().l2(1e-4).updater(new Adam(learningRate)).dataSetFeatureMapping("bertOutput").dataSetLabelMapping("label")
            .trainEvaluation("output2", 0, new Evaluation()).build();

        sd.setTrainingConfig(config);
      }

      CheckpointListener checkpointListener = CheckpointListener.builder(modelFolder).saveEveryNEpochs(1).deleteExisting(true).build();

      UIListener uiListener = UIListener.builder(new File(NerUtil.DATA_HOME + "/ner/ui/log.bin")).plotLosses(1).trainEvaluationMetrics("output2", 0, Evaluation.Metric.ACCURACY, Evaluation.Metric.F1)
          .updateRatios(1).trainEvalFrequency(10).build();

      PcngListener pcngListener = new PcngListener();

      sd.setListeners(uiListener, new ScoreListener(10), checkpointListener, pcngListener);

      VertxUIServer uiServer = VertxUIServer.getInstance();
      uiServer.start();
      // System.out.println(sd.summary());

      BertTrainIterator trainDataIter = new BertTrainIterator();
      trainDataIter.setTokenProvider(new TokenProvider());

      History history = sd.fit().train(trainDataIter, nEpochs).exec();
      List<Double> acc = history.trainingEval(Metric.ACCURACY);
      System.out.println("Accuracy: " + acc);
    } catch (Exception e) {
      e.printStackTrace();
    }
  }
}

**Comment by SidneyLann** [comment]
The losses should be smaller and smaller and not bigger and bigger in every epochs and iterations.

**Comment by SidneyLann** [comment]
![7](https://user-images.githubusercontent.com/5990314/71148664-3981b880-2267-11ea-8b19-42978c7bfe3f.jpg)
![8](https://user-images.githubusercontent.com/5990314/71148665-3981b880-2267-11ea-84fe-6eb2067d6d3b.jpg)

**Comment by SidneyLann** [comment]
https://ibb.co/4p7KPND
https://ibb.co/BccnCSX

**Comment by SidneyLann** [comment]
![image](https://user-images.githubusercontent.com/5990314/71151746-7b166180-226f-11ea-87a4-2661510df53b.png)



the loss2 is still zero at this line, and no chance to assign it after this line.

**Comment by SidneyLann** [comment]
This issue impacts the training result? or just impacts the UI?

**Comment by SidneyLann** [comment]
List<String> outputVars = new ArrayList<>(gradVarToVarMap.keySet());  
outputVars.addAll(lossVariables);//the loss2 has values now after added this line in TrainingSession.java, the result should be correct?
Map<String, INDArray> m = output(outputVars, placeholders, batch, requiredActivations, listeners, at);

**Comment by SidneyLann** [comment]
//Do training iteration
        List<String> outputVars = new ArrayList<>(gradVarToVarMap.keySet());    //TODO this should be empty, and grads calculated in requiredActivations
        //Sidney start
        outputVars.addAll(lossVariables); 
        //Sidney end
        Map<String, INDArray> m = output(outputVars, placeholders, batch, requiredActivations, listeners, at);




Need to add  outputVars.addAll(lossVariables);  in TrainingSession.java to have non-zero loss, is it normal?

**Comment by SidneyLann** [comment]
Not fixed in M1.

**Comment by agibsonccc** [comment]
@SidneyLann yeah we flagged relevant issues already. Thanks for mentioning this.

**Comment by SidneyLann** [comment]
Fixed in latest SNAPSHOT.

**Comment by CompilerCrash** [comment]
Hey @SidneyLann @agibsonccc, are you sure, this has been fixed and the fix is included in the latest snapshot?

When I try to fit my SameDiff graph, the reported loss (as per ScoreListener and LossCurve) is constantly zero and it learns nothing. This happened with different loss functions and CPU as well as on GPU. I created a minimal example [here](https://github.com/CompilerCrash/DL4J-Bug-Reproduction/blob/master/src/main/java/Loss.java) that should reproduce the situation. Could you please take a look at it? If I made no configuration mistake then either the problem persists or it's a new, similar one (then I can open a new issue for it).

**Comment by agibsonccc** [comment]
@CompilerCrash this was actually due to zero gradients. This has already been fixed . Open a new issue if you are not sure.
